### PR TITLE
dual gap limit

### PR DESCRIPTION
**Pull request by matbesancon**
termination should be checked with a duality gap of the whole tree instead of empty nodes.

### CODE REVIEW

File: src/Bonobo.jl [hunk 1]
```diff
@@ -300,7 +301,10 @@ end
 Return true when the branch and bound loop in [`optimize!`](@ref) should be terminated.
 Default behavior is to terminate the loop only when no nodes exist in the priority queue.
 """
-terminated(tree::BnBTree) = isempty(tree.nodes)
+function terminated(tree::BnBTree)
+    dual_gap = tree.incumbent - tree.lb
```
Comment (Wikunia) [inline src/Bonobo.jl:306]: Isn't it normally computed as a percentage?

File: src/Bonobo.jl [hunk 2]
```diff
@@ -300,7 +301,10 @@ end
 Return true when the branch and bound loop in [`optimize!`](@ref) should be terminated.
 Default behavior is to terminate the loop only when no nodes exist in the priority queue.
```
Comment (Wikunia) [inline src/Bonobo.jl:303]: would need to update the docstring here

**Review by Wikunia — CHANGES_REQUESTED** [review]

File: src/Bonobo.jl (hunk 1 above)
Comment (matbesancon) [reply src/Bonobo.jl:306]: you're right there are both, in SCIP:
```
# solving stops, if the relative gap = |primal - dual|/MIN(|dual|,|primal|) is below the given value, the gap is 'Infinity', if primal and dual bound have opposite signs
# [type: real, advanced: FALSE, range: [0,1.79769313486232e+308], default: 0]
limits/gap = 0

# solving stops, if the absolute gap = |primalbound - dualbound| is below the given value
# [type: real, advanced: FALSE, range: [0,1.79769313486232e+308], default: 0]
limits/absgap = 0
```

Comment (matbesancon) [reply src/Bonobo.jl]: fixed in last commit

File: src/Bonobo.jl (hunk 2 above)
Comment (matbesancon) [reply src/Bonobo.jl]: done

**Comment by matbesancon** [comment]
Now that I think of it, let's also add absolute gap

File: src/Bonobo.jl [hunk 3]
```diff
@@ -301,17 +303,18 @@ end
 
 Return true when the branch-and-bound loop in [`optimize!`](@ref) should be terminated.
 Default behavior is to terminate the loop only when no nodes exist in the priority queue
-or when the relative duality gap is below the tolerance fixed in the options.
+or when the relative or absolute duality gap are below the tolerances fixed in the options.
 """
 function terminated(tree::BnBTree)
+    absgap = tree.incumbent - tree.lb
```
[the remark this answers is not in the document]
Comment (matbesancon) [reply src/Bonobo.jl:309]: yes I assumed minimization here, in which case yup this is always in the same sense

**Comment by matbesancon** [comment]
all good on the comments it seems?

**Comment by Wikunia** [comment]
Looks good maybe would be nice to have some easy tests for this.

**Comment by matbesancon** [comment]
I added a simple test although the gap will only show up for large enough problems.

There was also a bug in a condition that was checking `:sense == :Max` instead of `sense == :Max`

**Comment by Wikunia** [comment]
Great thanks. Do you mind adding a line to the changelog then I'm happy to merge.

**Comment by matbesancon** [comment]
done!

**Pull request merged**